File: FutureVoice/Services/GoalStore.swift
import Foundation
import Combine

/// Daily challenge targets — how many word judgments / expression judgments /
/// shadow takes count as "done for today". Targets only; the reps themselves
/// come from `PracticeLog`. A target of 0 turns that challenge off.
///
/// UserDefaults-backed and global across languages: the habit is one habit,
/// whichever target language is active (PracticeLog is global for the same
/// reason).
@MainActor
final class GoalStore: ObservableObject {
    static let shared = GoalStore()

    @Published var sentencesPerDay: Int {
        didSet { defaults.set(sentencesPerDay, forKey: Keys.sentences) }
    }
    @Published var wordsPerDay: Int {
        didSet { defaults.set(wordsPerDay, forKey: Keys.words) }
    }
    @Published var expressionsPerDay: Int {
        didSet { defaults.set(expressionsPerDay, forKey: Keys.expressions) }
    }
    @Published var shadowsPerDay: Int {
        didSet { defaults.set(shadowsPerDay, forKey: Keys.shadows) }
    }

    private enum Keys {
        static let sentences   = "futurevoice.goal.sentencesPerDay"
        static let words       = "futurevoice.goal.wordsPerDay"
        static let expressions = "futurevoice.goal.expressionsPerDay"
        static let shadows     = "futurevoice.goal.shadowsPerDay"
    }

    private let defaults: UserDefaults

    init(defaults: UserDefaults = .standard) {
        self.defaults = defaults
        // 20 = one deck (`DrillView.sessionCap`), which is what the tile
        // asked for before this was settable.
        sentencesPerDay   = defaults.object(forKey: Keys.sentences) as? Int ?? 20
        wordsPerDay       = defaults.object(forKey: Keys.words) as? Int ?? 10
        expressionsPerDay = defaults.object(forKey: Keys.expressions) as? Int ?? 3
        shadowsPerDay     = defaults.object(forKey: Keys.shadows) as? Int ?? 2
    }

    var anyEnabled: Bool {
        sentencesPerDay > 0 || wordsPerDay > 0 || expressionsPerDay > 0 || shadowsPerDay > 0
    }

    /// Every enabled challenge met on `date`. False when nothing is enabled —
    /// a day with no goals can't be "met", or the streak would be infinite.
    func met(on date: Date, log: PracticeLog = .shared) -> Bool {
        guard anyEnabled else { return false }
        // FINISHED work only. Reps count every time an item was handled, so
        // reading those let a day complete itself by postponing cards.
        let day = log.day(date) ?? PracticeLog.Day()
        if sentencesPerDay > 0, day.drillDone < sentencesPerDay { return false }
        if wordsPerDay > 0, day.wordDone < wordsPerDay { return false }
        if expressionsPerDay > 0, day.expressionDone < expressionsPerDay { return false }
        if shadowsPerDay > 0, day.shadowDone < shadowsPerDay { return false }
        return true
    }

    /// Consecutive goal-met days ending today — or ending yesterday when today
    /// isn't done yet, so an unfinished morning doesn't read as a broken run.
    func streak(now: Date = Date(), calendar: Calendar = .current,
                log: PracticeLog = .shared) -> Int {
        guard anyEnabled else { return 0 }
        var day = now
        if !met(on: day, log: log) {
            guard let yesterday = calendar.date(byAdding: .day, value: -1, to: day) else { return 0 }
            day = yesterday
        }
        var count = 0
        while met(on: day, log: log) {
            count += 1
            guard let prev = calendar.date(byAdding: .day, value: -1, to: day) else { break }
            day = prev
        }
        return count
    }
}

File: android/app/src/main/java/com/roro/futurevoice/data/GoalStore.kt
package com.roro.futurevoice.data

import android.content.Context
import java.util.Calendar

/**
 * Daily challenge targets — how many sentence cards / word judgments /
 * expression judgments / shadow takes count as "done for today". Targets
 * only; the reps themselves come from [PracticeLog]. A target of 0 turns that
 * challenge off.
 *
 * Global across languages, like [PracticeLog]: the habit is one habit,
 * whichever target language happens to be active.
 */
object GoalStore {

    private const val PREFS = "futurevoice"
    private const val SENTENCES = "futurevoice.goal.sentencesPerDay"
    private const val WORDS = "futurevoice.goal.wordsPerDay"
    private const val EXPRESSIONS = "futurevoice.goal.expressionsPerDay"
    private const val SHADOWS = "futurevoice.goal.shadowsPerDay"

    data class Goals(
        /** 20 = one deck's session cap, which is what the tile asked for
         *  before this was settable. */
        val sentences: Int = 20,
        val words: Int = 10,
        val expressions: Int = 3,
        val shadows: Int = 2,
    ) {
        val anyEnabled: Boolean
            get() = sentences > 0 || words > 0 || expressions > 0 || shadows > 0
    }

    fun load(c: Context): Goals {
        val p = c.getSharedPreferences(PREFS, 0)
        return Goals(
            sentences = p.getInt(SENTENCES, 20),
            words = p.getInt(WORDS, 10),
            expressions = p.getInt(EXPRESSIONS, 3),
            shadows = p.getInt(SHADOWS, 2),
        )
    }

    fun save(c: Context, g: Goals) {
        c.getSharedPreferences(PREFS, 0).edit()
            .putInt(SENTENCES, g.sentences)
            .putInt(WORDS, g.words)
            .putInt(EXPRESSIONS, g.expressions)
            .putInt(SHADOWS, g.shadows)
            .apply()
    }

    /**
     * Every enabled challenge met on [at]. False when nothing is enabled — a
     * day with no goals cannot be "met", or the streak would be infinite.
     *
     * Reads FINISHED work only. Reps count every time an item was handled, so
     * reading those would let a day complete itself by postponing cards.
     */
    fun met(c: Context, g: Goals, at: Long): Boolean {
        if (!g.anyEnabled) return false
        val day = PracticeLog.day(c, at) ?: PracticeLog.Day()
        if (g.sentences > 0 && day.drillDone < g.sentences) return false
        if (g.words > 0 && day.wordDone < g.words) return false
        if (g.expressions > 0 && day.expressionDone < g.expressions) return false
        if (g.shadows > 0 && day.shadowDone < g.shadows) return false
        return true
    }

    /**
     * Consecutive goal-met days ending today — or ending YESTERDAY when today
     * isn't done yet, so an unfinished morning doesn't read as a broken run.
     */
    fun streak(c: Context, g: Goals = load(c), now: Long = System.currentTimeMillis()): Int {
        if (!g.anyEnabled) return 0
        val cal = Calendar.getInstance().apply { timeInMillis = now }
        if (!met(c, g, cal.timeInMillis)) cal.add(Calendar.DAY_OF_YEAR, -1)
        var count = 0
        while (met(c, g, cal.timeInMillis)) {
            count += 1
            cal.add(Calendar.DAY_OF_YEAR, -1)
        }
        return count
    }
}
